added transition between different pages (home/config/about) : Issue #74

vasusharma7 opened the pull request:
Fixes #74

I have implemented a transition effect between different pages . This is just a naive version. Will require enhancements. Please checkout the transition effect and guide further......

mossroy commented:
Cool, it's a starting point.
The transition works, but needs enhancement, like you said.

The goal is the user feels like the 3 sections are contiguous : clicking on the section names makes the visible zone move between these 3 sections.

```
-------------------
|home |conf |about|
|     |     |     |
|     |     |     |
-------------------
```

Currently, the transition is always from left to right : it should depend on the current section, in order to "move" to the other section in the right direction.
I also think the transition should be a bit faster

vasusharma7 commented:
Yes. I got it...I will implement the suggestions and get back here soon

vasusharma7 commented:
Sir, I have implemented the suggested changes...Please have a look at the transitions now...

vasusharma7 commented:
Also, I had  a doubt with the transitions that the transition from **home** to **about** should pass through **config** or directly to **about** as it is happening in current implementation ...

Jaifroid commented:
> Also, I had  a doubt with the transitions that the transition from home to about should pass through config or directly to about as it is happening in current implementation ...

Hi @vasusharma7, thanks for this. It's looking good! I don't think a direct switch from home to about should pass through config, because it adds overhead for very little gain.

On the code front, there's currently a fair amount of repetition in the added lines in `app.js`, so I think it would be better to tidy it all into a single function that can be called from each of the button click functions.

I think we should give the user the ability to turn off animations, because animations tend to render poorly on low-end devices. A simple option "Animate transition between app pages" under Display settings would be good, and it can default to on.

I suggest adding it using the new-style UI-setting code you'll find in app.js just below [line 78](https://github.com/kiwix/kiwix-js/blob/master/www/js/app.js#L78), and add the code to deal with the onclick event just below [line 293](https://github.com/kiwix/kiwix-js/blob/master/www/js/app.js#L293). You add the UI for the option in index.html after [line 241](https://github.com/kiwix/kiwix-js/blob/master/www/index.html#L241) - just copy the existing label for hide content active warning check, and edit.

In each case, simply copy the code for the corresponding `hideActiveContentWarningCheck` and place the new code immediately after. I'd probably name it `showUIAnimations` and `showUIAnimationsCheck`.

Jaifroid commented:
@vasusharma7 Just another couple of minor points:
1. Avoid using ES6 arrow function because we're not transpiling with Babel and have to support IE11 and Firefox OS;
2. It occurred to me that you could simplify the code quite a lot, and maybe make it generic, by using the pure JS `element.classList.toggle()` function -- see https://www.w3schools.com/howto/howto_js_toggle_class.asp . I think all our supported browsers can use `classList.toggle()`. Just a thought!

vasusharma7 commented:
Thanks for suggestions....I will implement all the changes soon and get back soon

mossroy commented:
The transition looks much better now.
I did not have a look at the code, but agree with the general remarks and suggestions of @Jaifroid 

Regarding the ability to turn off this feature, I don't find it critical.
I would be OK to have the transition feature (always on) in this PR, and add the on/off feature later with another PR, if you prefer.

vasusharma7 commented:
@Jaifroid @mossroy 
1. I have added a button to turn on/off the animation.
2. I have removed the arrow function from the code.                    
3. I also tried the toggle function but it was not fitting well maybe because sometimes classes have to be removed altogether and switching operation is not in use...I may be wrong also..
4. I have tried to remove the repitition as far as possible but still my code may be somewhat monotonous.......Please suggest more improvements and changes.
I couldn't bind the animation code in a single function because everytime a different set of statements have to be executed....

vasusharma7 commented:
Still have a look at the code and suggest some changes if any.....I am still a newbie.

Jaifroid commented on www/css/transition.css:
```diff
@@ -0,0 +1,69 @@
+/* COMMENT OUT ALL THE SCALE PARTS FOR A SMOOTH HORIZONTAL ANIMATION */
```
I guess you can delete this comment and the commented-out parts below?

vasusharma7 replied:
Ya ...  I will make this sure

Jaifroid commented on www/css/transition.css:
```diff
@@ -0,0 +1,69 @@
+/* COMMENT OUT ALL THE SCALE PARTS FOR A SMOOTH HORIZONTAL ANIMATION */
+.fadeOut_R{
```
Since we're not using a fade out, but a slide transition, can you rename these functions .slideOut_R and .slideInR, etc.?

Also, add spaces before all curly brackets here and below, e.g.: `.slideOut_R {`

vasusharma7 replied:
Ok .. I will do it

Jaifroid commented on www/js/app.js:
```diff
@@ -218,19 +220,88 @@ define(['jquery', 'zimArchiveLoader', 'util', 'uiUtil', 'cookies','abstractFiles
+        $('#about').removeClass("fadeIn_L").removeClass("fadeOut_R");
+        $('#configuration').removeClass("fadeIn_L").removeClass("fadeIn_R").removeClass("fadeOut_L").removeClass("fadeOut_R");
+        $('#articleContent').removeClass("fadeIn_R").removeClass("fadeOut_L");
+    }
```
I think this function should be renamed to `removeAnimationClasses` to be clear about what classes are being removed. I think you can easily move this function and the next one to `uiUtil.js` to help declutter `app.js`. Please also add spaces before open braces (curly brackets).

vasusharma7 replied:
Ok sure.... yes that makes more sense also

Jaifroid commented on www/js/app.js:
```diff
@@ -218,19 +220,88 @@ define(['jquery', 'zimArchiveLoader', 'util', 'uiUtil', 'cookies','abstractFiles
         // Show the selected content in the page
-        $('#about').hide();
-        $('#configuration').hide();
+            removeClasses();
```
Make sure these statements are aligned with existing code. They will become `uiUtil.removeClasses();` etc. if you move the functions as suggested earlier.

vasusharma7 replied:
Ok....I will implement the suggested soon

Jaifroid commented on www/js/app.js:
```diff
@@ -218,19 +220,88 @@ define(['jquery', 'zimArchiveLoader', 'util', 'uiUtil', 'cookies','abstractFiles
+        if(params.showUIAnimations){ 
+           addAnimation("home");
+        }
+        else{
```
Add spaces between if statements and parentheses, and also (as mentioned earlier) in front of opening braces.

vasusharma7 replied:
Ya ... I will make this sure

Jaifroid commented on www/js/app.js:
```diff
@@ -252,14 +323,19 @@ define(['jquery', 'zimArchiveLoader', 'util', 'uiUtil', 'cookies','abstractFiles
+        if(params.showUIAnimations){ 
+            addAnimation("config");
+        }
+        else{
```
Use our existing if … else style:

```
if (something) {
 ...
} else {
...
}
```

vasusharma7 replied:
Ok sure

Jaifroid commented on www/js/app.js:
```diff
@@ -890,7 +976,7 @@ define(['jquery', 'zimArchiveLoader', 'util', 'uiUtil', 'cookies','abstractFiles
             }
         }
-        ,500);
+        ,200);
```
Is there a reason for changing this? We actually need to slow this down for some slow browsers such as Chrome on Android...

vasusharma7 replied:
Actually it was asked in one of the earlier comments to make the speed faster...thats why I did this.....In my next commit, I will change this to 350 to be more optimum. You can see and suggest if it looks better

Jaifroid replied:
@vasusharma7 I'm a bit confused by this still. It doesn't relate to this PR, so if it should be changed, I think it should be in a separate PR, but maybe I'm missing something.

vasusharma7 replied:
> Cool, it's a starting point.
> The transition works, but needs enhancement, like you said.
> 
> The goal is the user feels like the 3 sections are contiguous : clicking on the section names makes the visible zone move between these 3 sections.
> 
> ```
> -------------------
> |home |conf |about|
> |     |     |     |
> |     |     |     |
> -------------------
> ```
> 
> Currently, the transition is always from left to right : it should depend on the current section, in order to "move" to the other section in the right direction.
> I also think the transition should be a bit faster


See the last line about the transition to be made faster........Thats why I changed it to 200.

mossroy replied:
There must be a misunderstanding here.
The number at this line is a delay between a keypress in the search field and starting a title search.
I also don't see how it is related to this PR.

What I was suggesting was to make the transition a bit faster between the sections. I suppose this could be done with the animation duration in CSS

vasusharma7 replied:
@mossroy @Jaifroid 
I deeply apologize for doing that.....That was my fault....I assure you not to repeat this again.
Actually I understood what you wanted me to do....While changing the animation speed, I accidentally also changed that 200 value....I have corrected and committed the corrected version....Please have a look.......I am sorry again

Jaifroid replied:
No worries! Easy to mix those up, and we all make mistakes, especially when learning about new code. We're very happy to have your development help.

vasusharma7 replied:
@Jaifroid 
Thanks for understanding....
Now what should i do next in this PR...
Please guide further....
I dont know if it is right to ask but is it now ready for merge....
If not...what all is left now..

vasusharma7 commented:
I have implemented nearly all suggested changes....Please have a look at all.
I am sorry if I may have missed a few.

Jaifroid commented on www/css/transition.css:
```diff
@@ -0,0 +1,53 @@
+@keyframes moveOutR {
+    0% {
+        transform : translateX(0%);
+
```
@vasusharma7 It's very minor, but remove this redundant blank line (line 12), and similar ones below (38, 48).

vasusharma7 replied:
OkI will do so

Jaifroid commented on www/js/app.js:
```diff
@@ -218,19 +220,79 @@ define(['jquery', 'zimArchiveLoader', 'util', 'uiUtil', 'cookies','abstractFiles
         return true;
     });
     // Top menu :
+    function addAnimation(section) {
```
@vasusharma7 Sorry I wasn't clear. I meant to move this function also to uiUtil.js.

vasusharma7 replied:
OK, I will move that also there

Jaifroid commented on www/js/app.js:
```diff
@@ -218,19 +220,79 @@ define(['jquery', 'zimArchiveLoader', 'util', 'uiUtil', 'cookies','abstractFiles
-        $('#about').hide();
-        $('#configuration').hide();
+        uiUtil.removeAnimationClasses();
+        if(params.showUIAnimations) { 
```
Missing space after `if` (to be consistent with our house style)

vasusharma7 replied:
Ok... I will make sure this in my next commit

Jaifroid commented on www/js/lib/uiUtil.js:
```diff
@@ -244,6 +252,7 @@ define([], function() {
         isElementInView: isElementInView,
-        htmlEscapeChars : htmlEscapeChars
+        htmlEscapeChars : htmlEscapeChars,
+        removeAnimationClasses : removeAnimationClasses,
```
@vasusharma7 The last exported function shouldn't have a comma. But as you'll probably add the `addAnimation` function here, you can just add that one after this (with no comma at the end of addAnimation).

vasusharma7 replied:
Ok i will make this sure

Jaifroid commented on www/js/lib/uiUtil.js:
```diff
@@ -244,6 +252,7 @@ define([], function() {
         displayFileDownloadAlert: displayFileDownloadAlert,
         isElementInView: isElementInView,
-        htmlEscapeChars : htmlEscapeChars
+        htmlEscapeChars : htmlEscapeChars,
```
@vasusharma7 Remove the extra space before the colon, like lines 248-254. And in the next line below.

vasusharma7 replied:
Ok sure

Jaifroid commented on www/js/app.js:
```diff
@@ -843,7 +918,7 @@ define(['jquery', 'zimArchiveLoader', 'util', 'uiUtil', 'cookies','abstractFiles
         request.onreadystatechange = function () {
             if (request.readyState === XMLHttpRequest.DONE) {
-                if ((request.status >= 200 && request.status < 300) || request.status === 0) {
+                if ((request.status >= 300 && request.status < 300) || request.status === 0) {
```
@vasusharma7 Did @mossroy specifically ask you to change this? That seems a bit strange, as this function is only used by testing and has nothing to do with this PR... ? Maybe a mistake?

vasusharma7 replied:
Yes....Its this is my mistake.....I will mend it next time

vasusharma7 commented:
@Jaifroid 
I have implemented the suggested changes and committed them.Please have a review.
As far as the speed of animation is concerned, please have a look at that also so that I can adjust accordingly

Jaifroid commented:
Hi @vasusharma7 I've just pushed some minor formatting changes as suggested by ESLint, for style consistency. I'll do a bit of testing and get back to you.

vasusharma7 commented:
Thanks @Jaifroid  for approval and you guidance also in the whole process.
Ok, I will wait for approval of @mossroy .

vasusharma7 commented:
@mossroy 
Please have a review on the changes committed by me in the PR.
Jaifroid has already approved the changes.
Please tell if the changes can be merged or they require some more detailing.

mossroy commented on www/js/lib/uiUtil.js:
```diff
@@ -227,12 +228,83 @@ define([], function() {
+     * @param {String} name_of_section It takes the name of the section to which the animation is to be added
+     * 
+     */
+    function addAnimation(section) {
```
I would suggest to rename this fonction to something like animateToSection, because it does not simply add classes

vasusharma7 replied:
Yes,I have done the renaming....that sounds more relative

Jaifroid replied:
@mossroy @vasusharma7 It's not very important, but `animateToSection` is odd English. Either: "animateSection", or "applyAnimationToSection", or "applyCSSAnimationToSection" would be more natural.

vasusharma7 replied:
Yes, applyAnimationToSection seems more appropirate.
I will change this in my next commit along with other changes.
I am waiting for @mossroy 's comments on the (https://github.com/kiwix/kiwix-js/pull/579#discussion_r365526418) conversation and will submit only a single commit.

mossroy commented on www/js/lib/uiUtil.js:
```diff
@@ -227,12 +228,83 @@ define([], function() {
+     * 
+     */
+    function addAnimation(section) {
+        if (section == 'home') {
```
The content of this function might be simplified and be more generic to make it easier to add another section. But it's not crucial, you can leave it this way if you don't see how to do it

vasusharma7 replied:
@mossroy 
I have made this more generic by adding switch statement .Now new sections can easily be added....Please have a look

vasusharma7 replied:
I am sorry that  I am not very clear about this...So please clarify.

vasusharma7 commented on www/js/lib/uiUtil.js:
```diff
@@ -227,12 +228,83 @@ define([], function() {
     }
+    
+    /**
+     * Adds the slide animation between deifferent sections
```
Yes,I have resolved it

mossroy commented on www/js/lib/uiUtil.js:
```diff
@@ -227,12 +228,88 @@ define([], function() {
+    /**
+     * Adds the slide animation between different sections
+     * 
+     * @param {String} name_of_section It takes the name of the section to which the animation is to be added
```
The param name is wrong here

vasusharma7 replied:
Should it be 'string' instead of 'name_of_section' 
Something like -
@param {String} string It takes the name of the section as string to which the animation is to be added

Jaifroid replied:
@vasusharma7 It should be `@param {String} section The name of the section to which the animation is to be added`

vasusharma7 replied:
Thanks @Jaifroid 
I was  a bit confused

mossroy commented on www/js/lib/uiUtil.js:
```diff
@@ -227,12 +228,88 @@ define([], function() {
+     */
+    function animateToSection(section) {
+        var sectionsMap = {'home':0,'config':1,'about':2}
+        switch(sectionsMap[section]) {
```
It's not what I was thinking about. Sorry I should have been more specific.
The 3 cases of this swithc are very similar and could be merged. Instead of using ${'configuration'}, you could use ${section} to show it, and use a loop on the other 2 sections, to hide them. Only one generic code to handle the 3 (or more) sections

vasusharma7 replied:
Yes , I have tried doing that but if you look at different switch cases closely, you will observe that each switch case has completely different set of statements and minor similarities, So it's not possible to change it.
Also one of the reasons is also that the sections are not just hidden, their classes are changed and that also every time different ones.

mossroy replied:
OK never mind.
So I'd rather revert to the previous version without the switch/case. The if/elseif/elseif looks more readable to me as you directly see the section name.

mossroy commented on www/js/app.js:
```diff
@@ -75,7 +75,9 @@ define(['jquery', 'zimArchiveLoader', 'util', 'uiUtil', 'cookies','abstractFiles
 
     // Set parameters and associated UI elements from cookie
     params['hideActiveContentWarning'] = cookies.getItem('hideActiveContentWarning') === 'true';
+    params['showUIAnimations'] = cookies.getItem('showUIAnimations') === 'true';
```
On first use (or, more generally, when the cookie is not set at all), this disables the animation feature.
My opinion is that this feature should be on by default (we still let the user disable it if he/she wants).

So, this line should be modified, so that, if the cookie does not exist, we set `params['showUIAnimations']` to true

vasusharma7 replied:
@mossroy 
Please review the changes I made in the recent commit.
I have implemented as suggested .

vasusharma7 commented:
Yes, I will squash the commits ....

Jaifroid commented:
I'll do a quick test of the latest code in the browsers and will report back.

vasusharma7 commented:
Ok @Jaifroid 
I will wait for your approval before squash

vasusharma7 commented:
@Jaifroid @mossroy 
Great thanks for your continuous support and guidance.
I have squashed the commits (There were a few merge conflicts with commit that Jaifroid committed, I don't know why, But I have resolved them.)
Happy to have contributed. Will keep doing so.
Hoping to have a merge soon...

vasusharma7 commented:
@Jaifroid @mossroy
Great thanks for your continuous support and guidance.
I have squashed the commits (There were a few merge conflicts with commit that Jaifroid committed, I don't know why, But I have resolved them.)
Happy to have contributed. Will keep doing so.
Hoping to have a merge soon...

vasusharma7 commented:
I still had some merge conflicts left after the squash/rebase. 
I have resolved the conflicts with a new commit but now I am not able to squash them together.
Can you help me in this regard....
I am also sorry at the same time that I am not able to do simple things efficiently.

vasusharma7 commented:
I tried a lot of things from my side but in vain .... 
It is reflected in logs that i have done some shabby things 
I am very sorry for that and also for not being able to resolve the conflict

Jaifroid commented:
Hi @vasusharma7 , I've fixed the commit issue you were having. It was probably caused by my having added a commit that you hadn't tracked or merged locally, so your repo was out of sync with the PR here. In any case, I've done a diff between your master and the PR I've just force-pushed, and they are identical. You will find your local copy is still out of sync, but don't worry about that. Just wait for @mossroy to merge.

vasusharma7 commented:
@Jaifroid 
Great thanks for your help.I was really stuck.
Yes. Eagerly waiting for the merge.
